include/trieste/regex.h

```diff
@@ -0,0 +1,100 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+#pragma once
+
+#include "source.h"
+
+#include <re2/re2.h>
+
+namespace trieste
+{
+  class REMatch
+  {
+  private:
+    Source source;
```

Comment:
Okay, so this is my understanding.  

* `source` is the full underlying string
* `sp` is a view onto the string.  The end of `sp` and the end of `source` are the same.

I wonder if this structure is really an `iterator` rather than `match` as it seems the start of `sp` is the current iteration point into `source`. 

It feels that `skip`, `consume` and `empty` are on the iterator API.

The fields `match`, `locations` and `matches` feel to me like a different type that is the return of `consume`.  Now you don't want to keep allocating them, but I almost wonder if 
```C++
class REIterator
{
   ...
   bool consume(const RE2& regex, REMatch& match)
   ...
};
```
would be a clearer interface.  As this means you don't need to keep doing the allocation, but you get a sense of the correct order of the API.

Probably not worth the change, but it is how I understood the code.